### DISCUSSION TITLE
Demo Session: How to measure the effectiveness of GitHub Copilot

### DISCUSSION THREAD

**Discussion by KittyChiu** [community/community#142105] [Other Feature Feedback, Questions, & Ideas]
![image](https://github.com/user-attachments/assets/7ccfeca0-e855-4b01-8a3a-5dce1ad4c4ea)

Developers using GitHub Copilot know its value, but how do you show that value to get continued buy-in from the business? How do you know if you're measuring what matters most, and how do you use those metrics to craft a story for executives? 

@mickeygousset and @kittychiu will be presenting this in the coming Universe'24. 
We would love to hear your thought here in this thread - so how do you measure your Copilot effectiveness?

Did you missed it? No worries, the recording is [here](https://youtu.be/QKkhS-yUF_g?si=SgoPayOj1Wa4PoM1).
 
## UPDATED 2024-10-30

Here're the link to the community's [github/copilot-survey-engine](https://github.com/github/copilot-survey-engine) which you can automate and trigger at every PR closure. For longer version of the developers survey, here's the [link](https://downloads.ctfassets.net/wfutmusr1t3h/66acuCKYqXme0aukY8Rn3x/a8c682946b0176db5860544ad85fffe7/2024-04-23-GitHub-CCI-LP-Copilot-Impact-Survey-NT-V003.pdf).

<details><summary>Sample code snippet to `calculate_org_deployment_frequency.sh`</summary>

```yml
#!/bin/bash
# 1. Log in via `gh auth login`
# 2. chmod +x calculate_org_deployment_frequency.sh
# 3. Run ./calculate_org_deployment_frequency.sh

# Variables
ORG="<YOUR_ORG>"

# Fetch repositories
repos=$(gh repo list $ORG --limit 1000 --json name -q '.[].name')

calculate_frequency() {
  local repo=$1
  deployments=$(gh api repos/$ORG/$repo/deployments)
  current_time=$(date +%s)
  last_period=$(($current_time - 15768000)) # 6 months in seconds

  successful_deployments=0

  for deployment in $(echo $deployments | jq -r '.[] | select(.created_at | fromdateiso8601 > '$last_period') | .id'); do
    statuses=$(gh api repos/$ORG/$repo/deployments/$deployment/statuses | jq -r '.[].state')
    for status in $statuses; do
      if [ "$status" == "success" ]; then
        successful_deployments=$((successful_deployments + 1))
        break
      fi
    done
  done

  echo "$repo: Total successful deployments in the last 6 months: $successful_deployments"
}

# Iterate over repositories and calculate deployment frequency
for repo in $repos; do
  calculate_frequency $repo
done

```
</details>

<details><summary>Sample code snippet to `calculate_avg_pr_per_developer.sh`</summary>

```yml
#!/bin/bash
# 1. Log in via `gh auth login`
# 2. chmod +x calculate_avg_pr_per_developer.sh
# 3. Run ./calculate_avg_pr_per_developer.sh

# Variables
ORG="<YOUR_ORG>"

# Fetch repositories
repos=$(gh repo list $ORG --limit 1000 --json name -q '.[].name')

calculate_average_pr_merged_per_user() {
  local repo=$1
  current_time=$(date +%s)
  start_date=$(date -u -v -1m +"%Y-%m-%dT%H:%M:%SZ") # 1 month

  prs=$(gh pr list --repo $ORG/$repo --limit 1000 --state merged --search "closed:>$start_date sort:updated-desc" --json mergedAt,mergedBy)

  declare user_pr_count
  user_pr_count=()
  for pr in $(echo $prs | jq -r '.[] | select(.mergedAt) | .mergedBy.login'); do
    user_pr_count[$pr]=$((user_pr_count[$pr] + 1))
  done

  total_users=${#user_pr_count[@]}
  total_prs=0

  for count in "${user_pr_count[@]}"; do
    total_prs=$((total_prs + count))
  done

  mean_pr_merged_per_user=0

  if [ $total_users -gt 0 ]; then
    mean_pr_merged_per_user=$(echo "scale=2; $total_prs / $total_users" | bc)
  fi

  echo "$repo: Average PRs merged per user in the last month period: $mean_pr_merged_per_user"
}

# Iterate over repositories and calculate average PR merged per developer
for repo in $repos; do
  calculate_average_pr_merged_per_user $repo
done

```
</details>

Lastly, we could not communicate to you earlier before the official announcement, the GA release of the Copilot Metrics API is now available. There are [new payload schema and API endpoint](https://docs.github.com/en/rest/copilot/copilot-metrics?apiVersion=2022-11-28#get-copilot-metrics-for-an-enterprise) to this GA release, do read this [changelog](https://github.blog/changelog/2024-10-30-github-copilot-metrics-api-ga-release-now-available) to find out more.

A sneak peek of noticeable differences between beta and GA payload schema,

![image](https://github.com/user-attachments/assets/36a2c6cf-c164-4820-8f07-e71a07c57be4)

## UPDATED 2024-11-27

![image](https://github.com/user-attachments/assets/163388ef-c5ec-4ba6-9bfd-83230bad8d54)

Here's the [PowerBI AppSource link](https://appsource.microsoft.com/en-us/product/power-bi/github.github-copilot-metrics) to fast track visualizing your Copilot activities. Check out this [gist](https://gist.github.com/ayodejiayodele/3115d60ce7420891d2f005f7395ce99c) on the setup instruction. Kudos to @ayodejiayodele and his team on making this available to the community.

**Comment by SatyamPote** [comment]
Measuring the effectiveness of GitHub Copilot involves several strategies to quantify its impact on productivity, code quality, and overall development efficiency. Here are some key approaches:

Code Quality Metrics: Assess improvements in code quality by tracking metrics like the number of bugs or issues in the codebase pre- and post-Copilot usage [[1](https://resources.github.com/learn/pathways/copilot/essentials/measuring-the-impact-of-github-copilot/)].

Productivity Analysis: Evaluate developer output by measuring the number of completed tasks, lines of code written, or features developed over a specific period [[3](https://resources.github.com/artificial-intelligence/how-to-measure-github-copilot-impact/)].

Efficiency Indicators: Monitor time spent on coding versus time saved by using Copilot's suggestions. Tools like velocity KPIs can help in this analysis [[6](https://www.faros.ai/blog/measure-the-benefits-of-github-copilot-best-practices)].

User Feedback: Collect qualitative feedback from developers about their experience with Copilot, focusing on ease of use and perceived value [[2](https://github.com/orgs/community/discussions/142105)].

Causal Analysis: If improvements are not evident, conduct a causal analysis to identify potential bottlenecks that may affect the perceived benefits [[6](https://www.faros.ai/blog/measure-the-benefits-of-github-copilot-best-practices)].

By combining these methods, organizations can effectively demonstrate Copilot's value and impact on their development processes.

**Comment by KittyChiu** [comment]
☝🏻 Check out the above update on the PowerBI dashboard.

**Comment by josecleto** [comment]
@KittyChiu 
Thanks for sharing.  Does this metrics works for the Azure DevOps platform as well?  If not, when will it be?

**Reply by KittyChiu** [reply]
@josecleto The metrics mentioned in the content are related to GitHub Copilot and GitHub APIs.
